src/Collapse/include/gudhi/Flag_complex_sparse_matrix.h

```diff
@@ -0,0 +1,376 @@
+    insert_vertex(u);
+    GUDHI_CHECK((u != v),
+            std::invalid_argument("Flag_complex_sparse_matrix::insert_new_edge with u == v"));
+
+    insert_vertex(v);
+#ifdef DEBUG_TRACES
+    std::cout << "Insertion of the edge begins " << u <<", " << v << std::endl;
+#endif  // DEBUG_TRACES
+
+    auto rw_u = vertex_to_row_.find(u);
+    auto rw_v = vertex_to_row_.find(v);
+#ifdef DEBUG_TRACES
+    std::cout << "Inserting the edge " << u <<", " << v << std::endl;
+#endif  // DEBUG_TRACES
+    sparse_row_adjacency_matrix_[rw_u->second].insert(rw_v->second) = idx;
+    sparse_row_adjacency_matrix_[rw_v->second].insert(rw_u->second) = idx;
+  }
+
+ public:
+  /** \brief Flag_complex_sparse_matrix constructor from a range of filtered edges.
+   *
+   * @param[in] filtered_edge_range Range of filtered edges. Filtered edges must be in
+   * `Flag_complex_sparse_matrix::Filtered_edge`.
+   *
+   * There is no need the range to be sorted, as it will be performed in
+   * `Flag_complex_sparse_matrix::filtered_edge_collapse`.
+   */
+  template<typename Filtered_edge_range>
+  Flag_complex_sparse_matrix(const Filtered_edge_range& filtered_edge_range)
+  : f_edge_vector_(filtered_edge_range.begin(), filtered_edge_range.end()) {
+    // To get the number of vertices
+    std::unordered_set<Vertex_handle> vertices;
+    for (Filtered_edge filtered_edge : filtered_edge_range) {
+      Vertex_handle u;
+      Vertex_handle v;
+      std::tie(u,v) = std::get<0>(filtered_edge);
+      vertices.emplace(u);
+      vertices.emplace(v);
+    }
+    // Initializing sparse_row_adjacency_matrix_, This is a row-major sparse matrix.
+    sparse_row_adjacency_matrix_.resize(vertices.size(), Sparse_vector(vertices.size()));
+  }
+
+  /** \brief Flag_complex_sparse_matrix constructor from a proximity graph, cf. `Gudhi::compute_proximity_graph`.
+   *
+   * @param[in] one_skeleton_graph The one skeleton graph. The graph must be in
+   * `Flag_complex_sparse_matrix::Proximity_graph`.
+   *
+   * The constructor is computing and filling a vector of `Flag_complex_sparse_matrix::Filtered_edge`
+   */
+  Flag_complex_sparse_matrix(const Proximity_graph& one_skeleton_graph)
+  : sparse_row_adjacency_matrix_(boost::num_vertices(one_skeleton_graph),
+                                 Sparse_vector(boost::num_vertices(one_skeleton_graph))) {
+    // Insert all edges
+    for (auto edge_it = boost::edges(one_skeleton_graph);
+         edge_it.first != edge_it.second; ++edge_it.first) {
+      auto edge = *(edge_it.first);
+      Vertex_handle u = source(edge, one_skeleton_graph);
+      Vertex_handle v = target(edge, one_skeleton_graph);
+      f_edge_vector_.push_back({{u, v}, boost::get(Gudhi::edge_filtration_t(), one_skeleton_graph, edge)});
+    }
+  }
+
+  /** \brief Performs edge collapse in a increasing sequence of the filtration value.
+   *
+   * \tparam FilteredEdgeOutput is a functor that furnishes `({Vertex_handle u, Vertex_handle v}, Filtration_value f)`
```

Comment:
Why not the flat (Vertex_handle u, Vertex_handle v, Filtration_value f)?

Reply:
Interface seems better : 7e92bb3a